### PR TITLE
[Lib][Vector] Add function for dot product between two sparse vectors

### PR DESCRIPTION
**Pull request by xwkuang5**
Add `T SparseVector<T>::dot_with_intcpt(const SparseVector<T>& b) const`
Add `T SparseVector<T>::dot(const SparseVector<T>& b) const`
Add `T SparseVector<T>::euclid_dist(const SparseVector<T>& b) const`

### CODE REVIEW

**Comment by xwkuang5** [comment]
I was implementing a machine learning program and I found out that currently there is no support for dot product between two sparse vectors. Did you leave them out because of some other issues? @Kelvin-Ng

**Comment by Kelvin-Ng** [comment]
I did not implement them because there is no efficient implementation... The most efficient way to implement it is to first convert one of them to dense vector and then do the dot product between dense and sparse vector. I am not sure if I should implement as I afraid that it gives a wrong impression to users that it is efficient. This is in fact the philosophy of C++ STL.

**Comment by Kelvin-Ng** [comment]
Oops, I did not notice that this is a pull request... I thought that this is an issue... I will read your implementation tomorrow.

**Comment by xwkuang5** [comment]
Oops, I mistakenly assume that the feature value pair will be in ascending order. I guess you are right. I will close this PR.

**Comment by Kelvin-Ng** [comment]
I see. Don't close it yet. Let me think about how to cope with this issue.

**Comment by Kelvin-Ng** [comment]
I think it is fine to put it there and put a comment above it saying that it can be used only when the elements are sorted.

Another way to do that is to add a flag `is_sorted` to record whether the elements are sorted. If not, it will sort it and then set the flag to true before doing the calculation.

@kygx-legend What do you think?

**Comment by Kelvin-Ng** [comment]
Ah... maybe use `sorted_dot()` as the name of the dot product function?

**Comment by xwkuang5** [comment]
But then the program can not be oblivious of the underlying representation of the vector, resulting in two separate pieces of codes for sparse vector and dense vector respectively.

**Comment by Kelvin-Ng** [comment]
We can also write `sorted_dot()` for dense vector which directly call `dot()`.

**Comment by kygx-legend** [comment]
I agree with using `sorted_dot()` and put a comment there for notice.

File: lib/vector.hpp [hunk 1]
```diff
@@ -202,12 +202,22 @@ class Vector<T, false> {
     T dot_with_intcpt(const DenseVector<T>&) const;
     T dot_with_intcpt(const SparseVector<T>&) const;
 
+    inline T sorted_dot(const DenseVector<T>& b) const { return (*this).dot(b); }
```
Comment (Kelvin-Ng) [inline lib/vector.hpp]: I think `dot(b)` is enough. Same below.

**Comment by Kelvin-Ng** [comment]
@xwkuang5 I think adding a `sort()` function is also useful. The `sort()` function can be just a blank function for dense vector.

File: lib/vector.hpp [hunk 2]
```diff
@@ -168,6 +168,8 @@ class Vector<T, false> {
 
     inline const T& operator[](int idx) const { return vec[idx]; }
 
+    inline void sortAsc() {}
```
Comment (Kelvin-Ng) [inline lib/vector.hpp]: Use `sort_asc()` for consistency.

File: lib/vector.hpp [hunk 3]
```diff
@@ -351,6 +362,10 @@ class Vector<T, true> {
 
     inline ConstFeaValIterator end_feaval() const { return vec.end(); }
 
+    inline void sortAsc() { std::sort(vec.begin(), vec.end(), [](const FeaValPair<T>& a, const FeaValPair<T>& b){
```
Comment (Kelvin-Ng) [inline lib/vector.hpp]: Use `sort_asc()` for consistency.

**Comment by Kelvin-Ng** [comment]
I think it is ready to merge after finishing the above minor changes. However, see issue #197

File: lib/vector.hpp (hunk 3 above)
Comment (kygx-legend) [inline lib/vector.hpp]: This line may not pass `clang-format.py`. Please check.

**Comment by xwkuang5** [comment]
I saw the discussion on issue #197. So the conclusion is that we will use Eigen in the future?

**Comment by Kelvin-Ng** [comment]
Yes, but we can still temporarily use our own `Vector`. We could gradually transit to Eigen.

If you need the dot product right now, I think you had better do the above minor modification and let us merge it.

**Pull request merged**